src/main/scala/uk/gov/nationalarchives/omega/api/services/ApiService.scala

```diff
@@ -0,0 +1,219 @@
+    val switched = state.compareAndSet(from, to)
+    if (!switched) {
+      val msg = s"Unable to switch ApiService state from: $from, to: $to."
+      if (errorOnInvalidServiceState) {
+        throw new IllegalStateException(msg)
+      } else {
+        logger.warn(msg)
+      }
+    }
+    switched
+  }
+
+  private def doStart(config: ServiceConfig): IO[ExitCode] = {
+
+    // TODO(AR) - one client, how to ack a consumer message after local persistence and then process it, and then produce a response
+    // TODO(AR) how to wire up queues and services using a config file or DSL?
+
+    // TODO(AR) request queue will typically be 1 (plus maybe a few more for expedited ops), response queues will be per external application
+
+    val localQueue: IO[Queue[IO, LocalMessage]] = Queue.bounded[IO, LocalMessage](config.maxLocalQueueSize)
+    val jmsConnector = new JmsConnector(config)
+
+    val result = for {
+      q <- localQueue
+      res <- jmsConnector.getJmsIo(QueueName(config.requestQueue)).use { case (jmsProducer, jmsConsumer) =>
+               process(q, jmsConsumer, jmsProducer)
+             }
+    } yield res
+    result.as(ExitCode.Success)
+  }
+
+  private def process(q: Queue[IO, LocalMessage], consumer: JmsAcknowledgerConsumer[IO], producer: JmsProducer[IO]) =
+    for {
+      // START
+      res <- IO.race(
+               transferMessageToLocalQueue(q)(consumer),
+
+               // single dispatcher
+               //        dispatcher(q).foreverM
+
+               // multiple dispatchers
+               List.range(start = 0, end = config.maxDispatchers).parTraverse_ { i =>
+                 IO.println(s"Starting consumer #${i + 1}") >>
+                   new Dispatcher(new LocalProducerImpl(producer, QueueName(config.replyQueue))).run(i)(q).foreverM
+               }
+             )
+      // END
+
+    } yield res
+
+  private def doStop(): Unit =
+    logger.info("Connection closed")
+
+  private def acknowledgeMessage(): IO[AckAction[IO]] =
+    //    IO.pure(AckAction.ack)
+    IO {
+      println("Acknowledged message")
+      AckAction.ack
+    }
+
+  private def createLocalMessage(persistentMessageId: PersistentMessageId, jmsMessage: JmsMessage): IO[LocalMessage] =
+    for {
+      sid       <- IO.fromOption(jmsMessage.getStringProperty("sid"))(throw new Exception("Missing service ID"))
+      serviceId <- IO.fromOption(ServiceIdentifier.withNameOption(sid))(throw new Exception("SID not recognised"))
+      messageId <- IO.fromOption(jmsMessage.getJMSMessageId)(throw new Exception("Missing message ID"))
+      text      <- jmsMessage.asTextF[IO]
+    } yield LocalMessage(persistentMessageId, serviceId, text, messageId)
+
+  // receives a message, persists it and then adds it to the queue
+  private def transferMessageToLocalQueue(
+    q: Queue[IO, LocalMessage]
+  )(jmsAcknowledgerConsumer: JmsAcknowledgerConsumer[IO]): IO[Unit] =
+    jmsAcknowledgerConsumer.handle { (jmsMessage, mf) =>
```

Comment:
Isn't this setting a response handler for a JMS queue? The method naming here implies it's being called once per message.

Reply:
True. Renamed to `createMessageHandler`.